

# Takeaway Orders


This data is based on an indian restaurant in UK .
Trying to find the taste and likes of combos for the people of UK.

Each row is a single product within the order.

There are currently ~200k rows of data.

In [1]:
import pandas as pd
import numpy as np

In [2]:
data=pd.read_csv('/content/restaurant-1-orders.csv')

In [3]:
data.head()

Order Number        Order Date  ... Product Price  Total products
0         16118  03/08/2019 20:25  ...          0.80               6
1         16118  03/08/2019 20:25  ...         12.95               6
2         16118  03/08/2019 20:25  ...          2.95               6
3         16118  03/08/2019 20:25  ...          3.95               6
4         16118  03/08/2019 20:25  ...          8.95               6

[5 rows x 6 columns]

In [4]:
#Let's change the format of the data

In [5]:
arr=data['Order Number'].values
d={}
for i in arr:
  if i not in d:
    d[i]=1
  else:
    d[i]+=1

In [6]:
len_uni=len(d.keys())
max_col=max(d.values())

In [7]:
max_col

60

In [8]:
all=[]


In [9]:
for i in d.keys():
  k=0
  x=data[data['Order Number']==i]['Item Name'].values
  x=list(x)
  all.append(x)


In [10]:
for i in range(len(all)):
  for j in range(60-len(all[i])):
    all[i].extend([float('NaN')])

In [11]:
# Let's make a new dataset now

In [12]:
dataset=pd.DataFrame(columns=[i for i in range(60)])

In [13]:
dataset=pd.DataFrame(all,columns=[i for i in range(60)])

In [14]:
# Let's check the newly made dataset
dataset.head()

0                   1                       2   ...   57   58   59
0   Plain Papadum    King Prawn Balti             Garlic Naan  ...  NaN  NaN  NaN
1      Plain Naan       Mushroom Rice  Tandoori Chicken (1/4)  ...  NaN  NaN  NaN
2      Aloo Chaat      Chicken Pakora   Chicken Chilli Garlic  ...  NaN  NaN  NaN
3  Chicken Pakora       Plain Papadum           Mango Chutney  ...  NaN  NaN  NaN
4   Chicken Balti  Special Fried Rice                     NaN  ...  NaN  NaN  NaN

[5 rows x 60 columns]

In [15]:
# Importing the apriori library and trying it

In [16]:
# Inspecting elements
all[:3]

[['Plain Papadum',
  'King Prawn Balti',
  'Garlic Naan',
  'Mushroom Rice',
  'Paneer Tikka Masala',
  'Mango Chutney',
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan],
 ['Plain Naan',
  'Mushroom Rice',
  'Tandoori Chicken (1/4)',
  'Vindaloo - Lamb',
  'Chapati',
  'Lamb Tikka',
  'Saag Paneer',
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan,
  nan]

In [19]:
transactions = []
for i in range(0, len(dataset)):
    transactions.append([str(dataset.values[i,j]) for j in range(0, 60)])

In [20]:
from apyori import apriori
rules = apriori(transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2)

In [21]:
# Visualising the results
results = list(rules)

In [22]:
lift = []
association = []
for i in range (0, len(results)):
    lift.append(results[:len(results)][i][2][0][3])
    association.append(list(results[:len(results)][i][0]))

In [23]:
# Visualizing the results in the dataframe

In [24]:
rank = pd.DataFrame([association, lift]).transpose()
rank.columns = ['Association', 'Lift']

In [25]:
# Show top 10 higher lift scores
rank.sort_values('Lift', ascending=False).head(10)

Association     Lift
71    [Cauliflower Bhajee, Chicken Tikka (Main), Saa...  48.2792
442   [Cauliflower Bhajee, nan, Chicken Tikka (Main)...  48.2792
339              [Plain Rice, Prawn Puree, Saag Bhajee]  36.8233
1025        [Plain Rice, Prawn Puree, nan, Saag Bhajee]  36.8233
42    [King Prawn Shaslick, Bombay Aloo, Chicken Sha...  36.6908
353   [nan, King Prawn Shaslick, Bombay Aloo, Chicke...  36.6908
7              [King Prawn Shaslick, Chicken Shashlick]  18.9759
91        [King Prawn Shaslick, nan, Chicken Shashlick]  18.9759
142           [Red Sauce, Dhansak Sauce, Onion Chutney]  18.7723
578   [Red Sauce, Mint Sauce, Dhansak Sauce, Onion C...  18.7723

Well from this we can see that the most of the people of the restraunt likes to order  Prawn puree nan saag and plain rice together 
. This might be the speciality of the restaurants.

**Eclat Implementation**

In [26]:
# All the unique items bought ..
for i in np.unique(data['Item Name'].values):
  print(i)

Aloo Chaat
Aloo Gobi
Aloo Methi
Baingan Hari Mirch
Bengal Fish Biryani
Bengal Fish Karahi
Bengal Fry Fish
Bengal King Prawn
Bengal Salad
Bhindi Bhajee
Bhuna
Bhuna - Chicken
Bhuna - Chicken Tikka
Bhuna - King Prawn
Bhuna - Lamb
Bhuna - Prawn
Bombay Aloo
Bottle Coke
Bottle Diet Coke
Brinjal Bhajee
Butter Chicken
COBRA ( LARGE )
COBRA (330ML)
COBRA (660ML)
Cauliflower Bhajee
Chana Masala
Chapati
Chicken Achar
Chicken Balti
Chicken Biryani
Chicken Chaat
Chicken Chaat Main
Chicken Chilli Garlic
Chicken Hari Mirch
Chicken Haryali
Chicken Karahi
Chicken Mysore
Chicken Pakora
Chicken Rezala
Chicken Roshni
Chicken Shashlick
Chicken Shashlick Curry
Chicken Sylhet
Chicken Tikka
Chicken Tikka (Main)
Chicken Tikka Balti
Chicken Tikka Biryani
Chicken Tikka Chilli Masala
Chicken Tikka Garlic
Chicken Tikka Jalfrezi
Chicken Tikka Jeera
Chicken Tikka Karahi
Chicken Tikka Masala
Chicken Tikka Pasanda
Chicken Tikka Sizzler
Coke 1.5 ltr
Curry
Curry - Chicken
Curry - Chicken Tikka
Curry - King Prawn
Curry -

**Creating the combinations with the pairs**

In [27]:
uniqueItems=np.unique(data['Item Name'].values)

In [28]:
pair = []
for j in range(0, len(uniqueItems)):
    k = 1;
    while k <= len(uniqueItems):
        try:
            pair.append([uniqueItems[j], uniqueItems[j+k]])
        except IndexError:
            pass
        k = k + 1;       

In [29]:
pair

[['Aloo Chaat', 'Aloo Gobi'],
 ['Aloo Chaat', 'Aloo Methi'],
 ['Aloo Chaat', 'Baingan Hari Mirch'],
 ['Aloo Chaat', 'Bengal Fish Biryani'],
 ['Aloo Chaat', 'Bengal Fish Karahi'],
 ['Aloo Chaat', 'Bengal Fry Fish'],
 ['Aloo Chaat', 'Bengal King Prawn'],
 ['Aloo Chaat', 'Bengal Salad'],
 ['Aloo Chaat', 'Bhindi Bhajee'],
 ['Aloo Chaat', 'Bhuna'],
 ['Aloo Chaat', 'Bhuna - Chicken'],
 ['Aloo Chaat', 'Bhuna - Chicken Tikka'],
 ['Aloo Chaat', 'Bhuna - King Prawn'],
 ['Aloo Chaat', 'Bhuna - Lamb'],
 ['Aloo Chaat', 'Bhuna - Prawn'],
 ['Aloo Chaat', 'Bombay Aloo'],
 ['Aloo Chaat', 'Bottle Coke'],
 ['Aloo Chaat', 'Bottle Diet Coke'],
 ['Aloo Chaat', 'Brinjal Bhajee'],
 ['Aloo Chaat', 'Butter Chicken'],
 ['Aloo Chaat', 'COBRA ( LARGE )'],
 ['Aloo Chaat', 'COBRA (330ML)'],
 ['Aloo Chaat', 'COBRA (660ML)'],
 ['Aloo Chaat', 'Cauliflower Bhajee'],
 ['Aloo Chaat', 'Chana Masala'],
 ['Aloo Chaat', 'Chapati'],
 ['Aloo Chaat', 'Chicken Achar'],
 ['Aloo Chaat', 'Chicken Balti'],
 ['Aloo Chaat', 'Chicken Bi

#### Calculating score
The calculation is done looking at the number of customers that bought both items (the pair) and divided by all customers of the week (7501). This calculation is done for all pairs possible and the score is returned on "score" list.

$ score = \frac{\text{number of lists that contain [item x and item y]}} {\text{number of all lists}} $

In [30]:
# number of the unique customers
l=len(d.keys())
print(l)

13397


In [31]:
%%time
score = []
for i in pair:
    cond = []
    for item in i:
        cond.append('("%s") in s' %item)
    mycode = ('[s for s in transactions if ' + ' and '.join(cond) + ']')
    #mycode = "print 'hello world'"
    score.append(len(eval(mycode))/l)

CPU times: user 5min 16s, sys: 97.9 ms, total: 5min 16s
Wall time: 5min 17s


In [32]:
# 10 most common orders for the takeaway
ranking_ECLAT = pd.DataFrame([pair, score]).transpose()
ranking_ECLAT.columns = ['Pair', 'Score']

In [33]:
ranking_ECLAT.sort_values('Score', ascending=False).head(10)

Pair      Score
28738         [Pilau Rice, Plain Papadum]   0.133239
28737            [Pilau Rice, Plain Naan]   0.129133
25315      [Mango Chutney, Plain Papadum]    0.11301
11651  [Chicken Tikka Masala, Pilau Rice]  0.0850937
28798         [Plain Naan, Plain Papadum]  0.0847951
16871           [Garlic Naan, Pilau Rice]  0.0814361
26197         [Mint Sauce, Plain Papadum]   0.080839
27245          [Onion Bhajee, Pilau Rice]  0.0794208
25313         [Mango Chutney, Pilau Rice]  0.0735239
28799            [Plain Naan, Plain Rice]  0.0681496

# Comparing the Eclat with the apriori

Well we can see that the item that attracts the customers is the pilau rice hoping this is one of the best rice from our indeian restaurant.
with the papadum which are the plain salted papad very famous dish and a part of indian cuisine.

we saw that both of them together had a score of 0.133 which is a pretty good score .


In [37]:
i = ['Pilau Rice', 'Plain Papadum']
cond = []
for item in i:
    cond.append('("{}") in s'.format(item))
mycode = ('[s for s in transactions if ' + ' and '.join(cond) + ']')
tra = eval(mycode)

In [38]:
cond

['("Pilau Rice") in s', '("Plain Papadum") in s']

In [39]:
mycode

'[s for s in transactions if ("Pilau Rice") in s and ("Plain Papadum") in s]'

In [40]:
tra

[['Chicken Pakora',
  'Plain Papadum',
  'Mango Chutney',
  'Red Sauce',
  'Korma - Prawn',
  'Peshwari Naan',
  'Pilau Rice',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['Chicken Tikka Masala',
  'Curry - Lamb',
  'Pilau Rice',
  'Peshwari Naan',
  'Plain Papadum',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',


In [45]:
print ('Score for "papdapum" and  "pilau rice": {}'.format(0.133239))

Score for "papdapum" and  "pilau rice": 0.133239
